linker/Mono.Linker.Steps/AddBypassNGenStep.cs

```diff
@@ -0,0 +1,133 @@
+			TypeDefinition bypassNGenAttribute = corelibMainModule.MetadataResolver.Resolve (bypassNGenAttributeRef);
+			MethodDefinition bypassNGenAttributeDefaultConstructor = null;
+
+			if (bypassNGenAttribute == null) {
+				// System.Runtime.BypassNGenAttribute is not found in corelib. Add it.
+				TypeReference systemAttributeRef = new TypeReference ("System", "Attribute", corelibMainModule, targetModule.TypeSystem.CoreLibrary);
+				TypeReference systemAttribute = corelibMainModule.MetadataResolver.Resolve (systemAttributeRef);
+				systemAttribute = corelibMainModule.ImportReference (systemAttribute);
+
+				if (systemAttribute == null)
+					throw new System.ApplicationException ("System.Attribute is not found in " + targetModule.TypeSystem.CoreLibrary.Name);
+
+				MethodReference systemAttributeDefaultConstructorRef = new MethodReference (".ctor", corelibMainModule.TypeSystem.Void, systemAttributeRef);
+				MethodReference systemAttributeDefaultConstructor = corelibMainModule.MetadataResolver.Resolve (systemAttributeDefaultConstructorRef);
+				systemAttributeDefaultConstructor = corelibMainModule.ImportReference (systemAttributeDefaultConstructor);
+
+				if (systemAttributeDefaultConstructor == null)
+					throw new System.ApplicationException ("System.Attribute has no default constructor");
+
+				bypassNGenAttribute = new TypeDefinition ("System.Runtime", "BypassNGenAttribute", TypeAttributes.NotPublic | TypeAttributes.Sealed, systemAttribute);
+
+				coreLibAssembly.MainModule.Types.Add (bypassNGenAttribute);
+
+				if (Annotations.GetAction (coreLibAssembly) == AssemblyAction.Copy) {
+					Annotations.SetAction (coreLibAssembly, AssemblyAction.Save);
+				}
+
+				var ctorAttributes = MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.SpecialName | MethodAttributes.RTSpecialName;
+				bypassNGenAttributeDefaultConstructor = new MethodDefinition (".ctor", ctorAttributes, coreLibAssembly.MainModule.TypeSystem.Void);
+				var instructions = bypassNGenAttributeDefaultConstructor.Body.Instructions;
+				instructions.Add (Instruction.Create (OpCodes.Ldarg_0));
+				instructions.Add (Instruction.Create (OpCodes.Call, systemAttributeDefaultConstructor));
+				instructions.Add (Instruction.Create (OpCodes.Ret));
+
+				bypassNGenAttribute.Methods.Add (bypassNGenAttributeDefaultConstructor);
+			}
+			else {
+				foreach (MethodDefinition method in bypassNGenAttribute.Methods) {
+					if (method.IsConstructor && !method.IsStatic && !method.HasParameters) {
+						bypassNGenAttributeDefaultConstructor = method;
+						break;
+					}
+				}
+
+				if (bypassNGenAttributeDefaultConstructor == null) {
+					throw new System.ApplicationException ("System.Runtime.BypassNGenAttribute has no default constructor");
+				}
+			}
+
+			MethodReference defaultConstructorReference = targetModule.ImportReference (bypassNGenAttributeDefaultConstructor);
```

Comment:
Did you test this with CoreLibrary uses as AddBypassNGen target ?

Reply:
Yes, I did. It works fine. In this case ImportReference simply returns the argument: https://github.com/mono/cecil/blob/master/Mono.Cecil/ModuleDefinition.cs#L876